pravega/bookkeeper-operator
# Issue 97: Providing upgrade path to a test version

Signed-off-by: SrishT <Srishti.Thakkar@dell.com>

### Change log description
This PR adds support to supply a custom upgrade path, which would enable testing the installation and upgrade of the bookkeeper cluster to non-released test versions.

### Purpose of the change
Fixes #97 

### What the code does
When the user enables `testmode` and provides both the `fromVersion` and `toVersion` inside the values.yaml, it does either of the following :
- if an entry corresponding to the fromVersion is already present in the bk-supported-versions configmap, the entry corresponding to the fromVersion in the bk-supported-versions configmap is replaced with a fresh entry which provides an upgrade path from fromVersion to toVersion
- if no entry corresponding to the fromVersion exists in the bk-supported-versions configmap, a new entry is created which provides an upgrade path from fromVersion to toVersion

Also when the user enables testmode and specifies just the fromVersion, a new entry is created in the bk-supported-versions configmap, which allows this version to be installed

### How to verify it
- Inside the values.yaml, set the value of testmode.enabled to true and provide values for testmode.fromVersion and testmode.toVersion. You should be able to see an entry of the following format added to the bk-supported-version configmap
```
<fromVersion> : <fromVersion> , <toVersion>
```
- set the value of testmode.enabled to true and provide a value only for testmode.fromVersion and leave testmode.toVersion as is. You should be able to see an entry of the following format added to the bk-supported-version configmap
```
<fromVersion> : <fromVersion>
```
- set the value of testmode.enabled to false. In this case you should not see any changes in the contents of the bk-supported-version configmap
